:rocket: AppVeyor updates and fixes

MisterDA opened the pull request:
This PR sports a good amount of updates and fixes for AppVeyor. Amongst the noticeable changes:
- move to Visual Studio 2022 by default;
- AppVeyor will only trigger on branch pushes which names match `4.**`, `5.**`, `trunk`, and all pull requests pushes, just as GitHub Actions;
- evict the `configure` cache if `configure` has changed (in a pull request), and always evict the configure cache on branch pushes (because it can't be determined whether `configure` has changed).

MisterDA committed:
AppVeyor: update to FlexDLL 0.43

MisterDA committed:
AppVeyor: don't pass build triplet to configure

It's not needed anymore and should have been x86_64-pc-cygwin.

MisterDA committed:
AppVeyor: default to Visual Studio 2022 environment

MisterDA committed:
AppVeyor: evict the configure cache if configure changed

The configure cache might be stale if the configure script has
changed: the script responsible for a cache entry might have been
modified. Detect whether configure has changed by computing a checksum
and retrieving the system version, and evict the cache if needed.

AppVeyor doesn't expose the commit ID of the previous state of the
branch on a branch push, so checking whether configure has changed by
inspecting the git history isn't possible.

MisterDA committed:
AppVeyor: set all git clones as a safe directory

MisterDA committed:
AppVeyor: restrict runs for branch pushes as for GitHub Actions

AppVeyor will still trigger on pull requests when the branch name
doesn't match the filter list, and will trigger on consecutive pushes
to an opened pull request.

dra27 commented on tools/ci/appveyor/appveyor_build.sh:
```diff
@@ -62,7 +62,7 @@ function set_configuration {
         ;;
         mingw64)
-            build='--build=i686-pc-cygwin'
+            build='--build=x86_64-pc-cygwin'
```
The others could be updated, too - although tbh there's no reason to be passing `--build` these days... perhaps just delete the build variable completely?

dra27 commented on tools/ci/appveyor/appveyor_build.sh:
```diff
@@ -95,9 +95,15 @@ function set_configuration {
 OCAMLROOT=$(echo "$OCAMLROOT" | cygpath -f - -m)
 
+# reset and set the git clone as safe
+git config --global --add safe.directory ''
```
It's a publicly hosted CI system - I think adding `*` here simply disables this "feature"!

dra27 commented on tools/ci/appveyor/appveyor_build.sh:
```diff
@@ -21,9 +21,7 @@ BUILD_PID=0
-  MAKE="make -j$NUMBER_OF_PROCESSORS"
-else
-  MAKE=make
+    export MAKEFLAGS="$MAKEFLAGS -j$NUMBER_OF_PROCESSORS"
```
No - this is potentially brittle and really surprising if one then writes `make` elsewhere expecting that to actually mean `make`! By all means rename the variable to be something clearer (e.g. `MAKE_PARALLEL` or some such) but I find exporting `MAKEFLAGS` to be in the same category as a certain opam distro's setenv on `CPP`...! 🙂

dra27 commented on tools/ci/appveyor/appveyor_build.sh:
```diff
@@ -20,9 +20,7 @@ BUILD_PID=0
-if [[ -z $APPVEYOR_PULL_REQUEST_HEAD_COMMIT ]] ; then
-    export MAKEFLAGS="$MAKEFLAGS -j$NUMBER_OF_PROCESSORS"
-fi
+export MAKEFLAGS="$MAKEFLAGS -j$NUMBER_OF_PROCESSORS"
```
FWIW, this was originally done to keep the log readable but I don't think that matters so much with the reduced console output!

dra27 commented on appveyor.yml:
```diff
@@ -12,6 +12,12 @@
 #*                                                                        *
 #**************************************************************************
 
+branches:
+  only:
+    - /4\.*/
+    - /5\.*/
+    - trunk
```
Does this _definitely_ work - I have a feeling that when [PR#11331](https://github.com/ocaml/ocaml/pull/11331) was created, there was an issue with trying to do the same for AppVeyor?

dra27 reviewed (commented):
I don't think you've over-done this 🙂

Apart from possibly... I really don't like the concept behind using `MAKEFLAGS` - it makes the calls to `make` very difficult to read (I'm anticipating future us looking at logs and briefly being puzzled why something ran in parallel or with unexpected flags).

I had a potentially better/alternate idea for the config cache - we could just use a checksum of `configure` and the version of Cygwin (from `uname`) as part of the filename - that should work reliably for both PRs and pushes?

nojb committed:
Fix regression in #12403: quote source code also for marshalled ASTs (#12698)

xavierleroy committed:
In C++, give `Bool_val` the type `bool` (#12702)

C++ 11 prohibits narrowing conversions from `int` to `bool` in some contexts
(list initialization).
C has no `bool` type by default, only `_Bool`, which is not in C++ by default.
So we end up with two definitions of `Bool_val`, one for C and one for C++.

Fixes: #11332
